### ISSUE TITLE
Don't show full text instruction in turn banner unless necessary

### ISSUE THREAD

**Issue by ericrwolfe** [mapbox/mapbox-navigation-ios#380]
Looking into a common issue where the turn banner shows a text instruction that is confusing and can lead to taking the wrong exit ramp.

In almost all cases, we should show the upcoming street/highway name/destination, and avoid showing a full text instruction until no other information is available. And event then, only when a maneuver is actionable (see #379).

![](https://user-images.githubusercontent.com/7937582/27878941-9a602bdc-6185-11e7-884d-49562f65dd04.png)

In this case, displaying the street/highway name in the banner with the turn icon would be sufficient, combined with a well-timed voice announcement. 

The directions response in the case above contains enough information to do this:
`https://api.mapbox.com/directions/v5/mapbox/driving-traffic/-97.81385339284961%2C30.258458222930003%3B-97.771627%2C30.227061.json?geometries=polyline&alternatives=true&steps=true&overview=full&access_token=<ADD TOKEN HERE>`
```json
{
    "maneuver": {
        "bearing_after": 118,
        "location": [
            -97.793362,
            30.232298
        ],
        "type": "merge",
        "bearing_before": 125,
        "modifier": "slight right",
        "instruction": "Merge right onto US 290"
    },
    "duration": 57.699999999999996,
    "distance": 1582.9,
    "name": "US 290; TX 71",
    "geometry": "{vowDng{sQxO{a@dC}Gz@oCl@aCd@cC^{BXeCNwBH_BLwEVcP",
    "ref": "US 290; TX 71",
    "weight": 57.699999999999996,
    "mode": "driving"
}
```

cc @bsudekum

**Comment by 1ec5** [comment]
First of all, I agree that we shouldn’t be showing the full, API-provided instruction text as far as ¾ mile away. That’s a regression.

The behavior of showing the full instruction text began as a way to avoid displaying something silly like “0 ft” at an intersection. But we attached it to the low alert level, the threshold for which has steadily crept away from the maneuver zone. We should detach the display of a full instruction from the alert level concept and display it only when we’re confident the user is within the junction (whether an intersection or highway exit).

This behavior has nothing to do with whether we have other information about the maneuver. We could generate our own instruction from structured data if we wanted to. In fact, we should either do that or start sending the API a `language` (mapbox/MapboxDirections.swift#133), because currently an English instruction like this appears regardless of the locale.

**Comment by 1ec5** [comment]
> Merging right takes you to west…

This comment points to a right-hand off-ramp. That would be described as “take the ramp” or at least “keep right at the fork”, not “merge right”, but I can understand why that would be confusing while driving.

The instruction here actually refers to [this point](http://www.openstreetmap.org/node/152372582) where US 290 merges into Loop 360 from the left. Since the user is currently on Loop 360, they would technically need to merge right onto US 290.

Speaking of which: the Loop 360 shield is visible, which is also a bug, since the user is currently on Loop 360, not heading toward it. That bug should now be fixed as of #353 and v0.5.0. Now the user will see the US 290 shield instead.

**Comment by ericrwolfe** [comment]
Our improved destination handling (https://github.com/mapbox/mapbox-navigation-ios/pull/417) should handle the core issue here: that the turn banner shows a verbose, prematurely actionable text instruction that is confusing and can lead to taking the wrong exit ramp.
